pyqtorch/modules/circuit.py

```diff
@@ -43,15 +43,26 @@ def __init__(self, n_qubits: int, operations: list):
 
     def __mul__(self, other: AbstractGate | QuantumCircuit) -> QuantumCircuit:
         if isinstance(other, QuantumCircuit):
-            return QuantumCircuit(
-                max(self.n_qubits, other.n_qubits), self.operations.extend(other.operations)
-            )
+            n_qubits = max(self.n_qubits, other.n_qubits)
+            return QuantumCircuit(n_qubits, self.operations.extend(other.operations))
 
         if isinstance(other, AbstractGate):
-            return QuantumCircuit(max(self.n_qubits, other.n_qubits), self.operations.append(other))
+            n_qubits = max(self.n_qubits, other.n_qubits)
+            return QuantumCircuit(n_qubits, self.operations.append(other))
 
         else:
-            return NotImplemented
+            return ValueError(f"Cannot compose {type(self)} with {type(other)}")
+
+    def __key(self) -> tuple:
+        return (self.n_qubits, *self.operations)
+
+    def __eq__(self, other: Any) -> bool:
+        if isinstance(other, QuantumCircuit):
+            return self.__key() == other.__key()
+        return NotImplemented
+
+    def __hash__(self) -> int:
+        return hash(self.__key())
```

Comment:
the only problem with this in our case is that changing the `operations: ModuleList` is going to change the hash which is not what python/pytorch is expecting... pytorch is returning a different hash for every created object, so also different for `ModuleList`s that contain the same elements. we could go for `hash() = super().__hash__()` or something, but then `__eq__` and `__hash__` have inconsistent behaviour... not sure which way to go

Reply:
i would rather go for just hashing the names of the gates in the module list. this should return true if the circuit is the same right? not necessarily if the (random) param values are the same

Reply:
I am not sure if I understand what you mean by names of the gates... for e.g. `X([0],1)` just hash `X`? then `X([1],2)` would have the same hash...